# Carga de datos y primer vistazo

In [8]:
import pandas as pd

# 1. Cargamos el dataset
df = pd.read_csv("./datasets/COVID.csv", index_col=0)

# 2. Vemos el tamaño del dataset (Filas, Columnas)
print(f"El dataset tiene {df.shape[0]} filas y {df.shape[1]} columnas.\n")



El dataset tiene 60160 filas y 21 columnas.



In [9]:
# 3. Inspeccionamos las primeras 3 filas
print("--- Primeras filas del dataset ---")
df.head(3)

--- Primeras filas del dataset ---


,Tweet URL,Tweet Posted Time (UTC),Tweet Content,Tweet Type,Client,Retweets Received,Likes Received,Tweet Location,Lat,Long,...,User Id,Name,Screen Name,User Bio,Verified or Non-Verified,Profile URL,Protected or Non-protected,User Followers,User Following,User Account Creation Date
Tweet Id,,,,,,,,,,,,,,,,,,,,,
"""1233417783175778304""",https://twitter.com/Giussi92/status/1233417783...,28 Feb 2020 15:44:49,Also the entire Swiss Football League is on ho...,Tweet,Twitter for iPhone,0,0,NaN,NaN,NaN,...,"""1556856595""",Giuseppe Gentile,Giussi92,NaN,Verified,https://twitter.com/Giussi92,Non-Protected,3071,100,30 Jun 2013 00:27:50
"""1233417742520332290""",https://twitter.com/LAMofficial/status/1233417...,28 Feb 2020 15:44:40,World Health Org Official: Trump’s press confe...,Tweet,Twitter Web App,0,0,Los Angeles CA USA,NaN,NaN,...,"""113738369""",London After Midnight,LAMofficial,London After Midnight is a music project by wr...,Verified,https://twitter.com/LAMofficial,Non-Protected,4189,84,12 Feb 2010 21:43:17
"""1233417741027225602""",https://twitter.com/mitchellvii/status/1233417...,28 Feb 2020 15:44:39,"I mean, Liberals are cheer-leading this #Coron...",Tweet,Twitter Web App,23,64,"Miami, FL",NaN,NaN,...,"""17980523""",Bill Mitchell,mitchellvii,Host of YourVoice™ America at http://yourvoice...,Verified,https://twitter.com/mitchellvii,Non-Protected,534045,10037,09 Dec 2008 01:54:21


El dataset tiene 60,160 tweets (filas) y 22 columnas. Es una muestra robusta, idónea para sacar patrones.

# Análisis de la estructura (Tipos de datos y nulos)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 60160 entries, "1233417783175778304" to "1200977067266990080"
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Tweet URL                   60160 non-null  object 
 1   Tweet Posted Time (UTC)     60160 non-null  object 
 2   Tweet Content               60160 non-null  object 
 3   Tweet Type                  60160 non-null  object 
 4   Client                      60160 non-null  object 
 5   Retweets Received           60160 non-null  int64  
 6   Likes Received              60160 non-null  int64  
 7   Tweet Location              40620 non-null  object 
 8   Lat                         19 non-null     float64
 9   Long                        19 non-null     float64
 10  Tweet Language              59012 non-null  object 
 11  User Id                     60160 non-null  object 
 12  Name                        60159 non-null  object 
 13  

In [11]:
df.isnull().sum()

Tweet URL                         0
Tweet Posted Time (UTC)           0
Tweet Content                     0
Tweet Type                        0
Client                            0
Retweets Received                 0
Likes Received                    0
Tweet Location                19540
Lat                           60141
Long                          60141
Tweet Language                 1148
User Id                           0
Name                              1
Screen Name                       0
User Bio                      10442
Verified or Non-Verified          0
Profile URL                       0
Protected or Non-protected        0
User Followers                    0
User Following                    0
User Account Creation Date        0
dtype: int64

Las columnas de coordenadas (Lat y Long) están vacías casi al 100% (solo hay 19 tweets con coordenadas de 60,000). Conclusión de negocio: Queda totalmente descartado hacer mapas de calor precisos por coordenadas en Power BI; nos enfocaremos en la columna de texto libre

In [12]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Retweets Received,60160.0,3.142886,106.443453,0.000000,0.000000,0.000000,0.00000,2.383200e+04
Likes Received,60160.0,5.272590,98.297300,0.000000,0.000000,0.000000,0.00000,1.752600e+04
Lat,19.0,-45.983002,91.125375,-122.395853,-111.911088,-82.616026,-31.06970,1.397000e+02
Long,19.0,32.120267,13.136373,-11.000000,33.947780,35.691324,39.68805,4.085376e+01
User Followers,60160.0,92749.349900,830970.664927,0.000000,145.000000,550.000000,2449.50000,2.127516e+07
User Following,60160.0,2193.177809,11963.820775,0.000000,204.000000,562.000000,1692.00000,1.182667e+06


El 50% de los tweets del dataset tienen 0 retweets y 0 likes, pero el máximo de retweets es de 23,832. Conclusión: Hay unos pocos usuarios con un impacto masivo (Outliers/Influencers).

In [13]:
df.describe(include='O').T

,count,unique,top,freq
Tweet URL,60160,60160,https://twitter.com/Azeria64Azeria/status/1205...,1
Tweet Posted Time (UTC),60160,19450,31 Jan 2020 22:30:00,29
Tweet Content,60160,24943,RT @SaludPublicaEs: El Centro Nacional de Micr...,1943
Tweet Type,60160,3,ReTweet,45291
Client,60160,315,Twitter for Android,23250
Tweet Location,40620,14538,United States,702
Tweet Language,59012,43,English,33174
User Id,60160,43297,"""306127388""",249
Name,60159,41540,All Express News,249
Screen Name,60160,43297,bitcoinconnect,249


# Exploración de variables clave

In [28]:
df['Tweet Language'].unique()

array(['English', 'Spanish', 'Japanese', 'Portuguese', 'Turkish',
       'French', 'Thai', 'German', 'Chinese', 'Desconocido', 'in',
       'Italian', 'Finnish', 'Urdu', 'Catalan', 'Arabic', 'Tagalog',
       'Polish', 'Korean', 'Hindi', 'Swedish', 'Estonian', 'Dutch', 'km',
       'iw', 'Tamil', 'Greek', 'Farsi', 'si', 'Pashto', 'Russian',
       'Romanian', 'Czech', 'Lithuanian', 'Basque', 'Welsh', 'ht',
       'Danish', 'Vietnamese', 'Latvian', 'ml', 'Divehi', 'Slovenian',
       'or'], dtype=object)

In [29]:
# Exploración de variables clave 
print("--- Distribución por Idioma (Top 5) ---")
print(df['Tweet Language'].value_counts())

print("\n--- Distribución de Cuentas Verificadas ---")
print(df['Verified or Non-Verified'].value_counts(normalize=True) * 100)

--- Distribución por Idioma (Top 5) ---
Tweet Language
English        33174
Spanish        15814
Italian         2958
French          2932
Desconocido     1148
Portuguese       912
Turkish          884
German           434
in               388
Chinese          283
Thai             271
Japanese         204
Tagalog          131
Catalan          104
Dutch             93
Arabic            79
Polish            69
Urdu              53
Russian           29
Greek             24
Danish            19
Korean            18
Farsi             18
Swedish           17
Hindi             15
Finnish           14
Tamil             13
Estonian          12
ht                 9
Czech              9
Romanian           8
Vietnamese         3
Slovenian          3
si                 3
ml                 3
Latvian            2
Lithuanian         2
Welsh              2
iw                 1
km                 1
Basque             1
Pashto             1
Divehi             1
or                 1
Name: count, dtype: i

La conversación se divide principalmente entre Inglés (33,174 tweets) y Español (15,814 tweets). Los demás idiomas (italiano, francés...) se quedan muy por detrás. Conclusión: Podríamos segmentar el dashboard final en estos dos grandes idiomas para que las métricas no se diluyan.

El 90.1% de las cuentas no están verificadas. La conversación sobre el COVID-19 en este set de datos está empujada por la ciudadanía general, no por grandes corporaciones o instituciones verificadas.

# Limpieza de datos

In [15]:
df = df.drop_duplicates()

In [16]:
## eliminar columnas que no aportan valor analítico
columns_to_drop = ['Lat','Long']
df = df.drop(columns=columns_to_drop)   

# Gestión de Nulos

In [ ]:
# Rellenamos los nulos con etiquetas genéricas
df['Tweet Location'] = df['Tweet Location'].fillna('Desconocido')
df['Tweet Language'] = df['Tweet Language'].fillna('Desconocido')

## podria generado otras opciones


df['Verified or Non-Verified'] = df['Verified or Non-Verified'].fillna('Desconocido')
df['Name'] = df['Name'].fillna('Desconocido')   

In [18]:
# 🛠️ CORRECCIÓN DE TIPOS NUMÉRICOS
df['Retweets Received'] = pd.to_numeric(df['Retweets Received'], errors='coerce')
df['Likes Received'] = pd.to_numeric(df['Likes Received'], errors='coerce')
df['User Followers'] = pd.to_numeric(df['User Followers'], errors='coerce').fillna(0).astype(int)
# Rellenamos los posibles valores corruptos (que ahora son NaN) con 0 interacciones
df['Retweets Received'] = df['Retweets Received'].fillna(0).astype(int)
df['Likes Received'] = df['Likes Received'].fillna(0).astype(int)
df['User Followers'] = df['User Followers'].fillna(0).astype(int)

In [19]:
#La columna de fecha viene como texto. La convertimos a formato datetime real.
df['Tweet Posted Time (UTC)'] = pd.to_datetime(df['Tweet Posted Time (UTC)'], errors='coerce')

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 60160 entries, "1233417783175778304" to "1200977067266990080"
Data columns (total 26 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Tweet URL                   60160 non-null  object        
 1   Tweet Posted Time (UTC)     60160 non-null  datetime64[ns]
 2   Tweet Content               60160 non-null  object        
 3   Tweet Type                  60160 non-null  object        
 4   Client                      60160 non-null  object        
 5   Retweets Received           60160 non-null  int64         
 6   Likes Received              60160 non-null  int64         
 7   Tweet Location              60160 non-null  object        
 8   Tweet Language              60160 non-null  object        
 9   User Id                     60160 non-null  object        
 10  Name                        60160 non-null  object        
 11  Screen Name            

In [33]:
# Creamos columnas derivadas para facilitar la segmentación en el Dashboard
df['Año-Mes'] = df['Tweet Posted Time (UTC)'].dt.to_period('M').astype(str)
df['Hora del Día'] = df['Tweet Posted Time (UTC)'].dt.hour

In [34]:
df.head()

,Tweet URL,Tweet Posted Time (UTC),Tweet Content,Tweet Type,Client,Retweets Received,Likes Received,Tweet Location,Tweet Language,User Id,...,User Followers,User Following,User Account Creation Date,Año-Mes,Hora del Día,Pais_Estandarizado,Categoria_Influencia,Tiene_Mencion,Tiene_Enlace,Dispositivo_Limpio
Tweet Id,,,,,,,,,,,,,,,,,,,,,
"""1233417783175778304""",https://twitter.com/Giussi92/status/1233417783...,2020-02-28 15:44:49,Also the entire Swiss Football League is on ho...,Tweet,Twitter for iPhone,0,0,Desconocido,English,"""1556856595""",...,3071,100,30 Jun 2013 00:27:50,2020-02,15,No Especificado,Usuario Activo (501-5k),No,Sí,Apple iOS
"""1233417742520332290""",https://twitter.com/LAMofficial/status/1233417...,2020-02-28 15:44:40,World Health Org Official: Trump’s press confe...,Tweet,Twitter Web App,0,0,Los Angeles CA USA,English,"""113738369""",...,4189,84,12 Feb 2010 21:43:17,2020-02,15,Estados Unidos,Usuario Activo (501-5k),No,Sí,Web (Escritorio)
"""1233417741027225602""",https://twitter.com/mitchellvii/status/1233417...,2020-02-28 15:44:39,"I mean, Liberals are cheer-leading this #Coron...",Tweet,Twitter Web App,23,64,"Miami, FL",English,"""17980523""",...,534045,10037,09 Dec 2008 01:54:21,2020-02,15,Otros / No Identificado,Celebridad / Institución (>50k),No,No,Web (Escritorio)
"""1233417699264356357""",https://twitter.com/HelenKennedy/status/123341...,2020-02-28 15:44:29,"Under repeated questioning, Pompeo refuses to ...",Tweet,Twitter for iPhone,4,11,NYC and the North Fork,English,"""2199541""",...,27704,1060,25 Mar 2007 19:14:46,2020-02,15,Estados Unidos,Micro-Influencer (5k-50k),No,Sí,Apple iOS
"""1233417674274807808""",https://twitter.com/W7VOA/status/1233417674274...,2020-02-28 15:44:23,#coronavirus comments now from @larry_kudlow h...,Tweet,Twitter for iPhone,0,0,James S. Brady Press Briefing Room,English,"""17919393""",...,94269,1999,06 Dec 2008 08:56:45,2020-02,15,Otros / No Identificado,Celebridad / Institución (>50k),Sí,Sí,Apple iOS


In [21]:
# Aseguramos que los nombres de las categorías estén estandarizados
df['Verified or Non-Verified'] = df['Verified or Non-Verified'].str.strip()

In [22]:
# la columna geográfica es uno de los mayores desafíos (y oportunidades) en esta prueba técnica.
# Si miramos la columna Tweet Location, verás que es un campo de texto libre. Eso significa que un usuario escribe "Madrid, Spain", otro "BCN", otro "En mi casa" y muchos lo dejan vacío

# El objetivo de este paso es crear una nueva columna limpia llamada Pais_Estandarizado mediante técnicas de búsqueda de palabras clave (Keywords).


# Rellenamos los nulos para poder aplicar funciones de texto sin que de error
df['Tweet Location'] = df['Tweet Location'].fillna('Desconocido').astype(str)

def extraer_pais(texto):
    texto_min = texto.lower() # Convertimos a minúsculas para facilitar la búsqueda
    
    # Mapeo de palabras clave a Países Estandarizados
    if 'spain' in texto_min or 'españa' in texto_min or 'madrid' in texto_min or 'barcelona' in texto_min:
        return 'España'
    elif 'usa' in texto_min or 'united states' in texto_min or 'estados unidos' in texto_min or 'california' in texto_min or 'ny' in texto_min:
        return 'Estados Unidos'
    elif 'uk' in texto_min or 'united kingdom' in texto_min or 'reino unido' in texto_min or 'london' in texto_min or 'londres' in texto_min:
        return 'Reino Unido'
    elif 'mexico' in texto_min or 'méxico' in texto_min or 'cdmx' in texto_min:
        return 'México'
    elif 'argentina' in texto_min or 'buenos aires' in texto_min:
        return 'Argentina'
    elif 'colombia' in texto_min or 'bogota' in texto_min or 'bogotá' in texto_min:
        return 'Colombia'
    elif 'italy' in texto_min or 'italia' in texto_min or 'roma' in texto_min:
        return 'Italia'
    elif 'france' in texto_min or 'francia' in texto_min or 'paris' in texto_min:
        return 'Francia'
    elif 'desconocido' in texto_min:
        return 'No Especificado'
    else:
        return 'Otros / No Identificado' # Para ubicaciones raras como "En mi nube" o países con pocos tweets

# 3. Aplicamos la función fila por fila creando la nueva columna
df['Pais_Estandarizado'] = df['Tweet Location'].apply(extraer_pais)

# 4. Revisamos rápidamente cómo quedó la distribución
print("--- Distribución de Tweets por País Estandarizado ---")
print(df['Pais_Estandarizado'].value_counts())

--- Distribución de Tweets por País Estandarizado ---
Pais_Estandarizado
Otros / No Identificado    29161
No Especificado            19541
Estados Unidos              4058
España                      2528
Reino Unido                 1215
Francia                     1124
México                       979
Italia                       799
Argentina                    447
Colombia                     308
Name: count, dtype: int64


Agrupar a los usuarios en "Categorías de Influencia" según sus seguidores.

In [23]:
#Nos aseguramos de que las columnas críticas sean estrictamente numéricas
df['User Followers'] = pd.to_numeric(df['User Followers'], errors='coerce').fillna(0).astype(int)
df['Retweets Received'] = pd.to_numeric(df['Retweets Received'], errors='coerce').fillna(0).astype(int)

def clasificar_influencia(seguidores):
    seg = int(seguidores)
    if seguidores < 500:
        return 'Micro-usuario (0-500)'
    elif seguidores < 5000:
        return 'Usuario Activo (501-5k)'
    elif seguidores < 50000:
        return 'Micro-Influencer (5k-50k)'
    else:
        return 'Celebridad / Institución (>50k)'

df['Categoria_Influencia'] = df['User Followers'].apply(clasificar_influencia)

En Twitter es crucial saber si el tuit es una opinión aislada o si intenta interactuar. Viendo el contenido de Tweet Content, muchos tuits incluyen links (https://...) o mencionan a la Organización Mundial de la Salud (@WHO) o a políticos (@Trump).

In [24]:
# Verificamos si el tuit menciona a alguien (contiene '@')
df['Tiene_Mencion'] = df['Tweet Content'].str.contains('@', na=False).map({True: 'Sí', False: 'No'})

# Verificamos si comparte un enlace externo (contiene 'http')
df['Tiene_Enlace'] = df['Tweet Content'].str.contains('http', na=False).map({True: 'Sí', False: 'No'})

La columna Client nos dice desde dónde se publicó el tuit (Twitter for iPhone, Twitter Web App, TweetDeck, etc.). Hay cientos de variantes.

In [25]:
df['Client'].unique()

array(['Twitter for iPhone', 'Twitter Web App', 'SocialFlow', 'TweetDeck',
       'Twitter Media Studio - LiveCut', 'Twitter for Android',
       'Twitter Web Client', 'Twitter for Advertisers', 'Periscope',
       'Buffer', 'SocialNewsDesk', 'Echobox Social', 'POZ Auto Tweet',
       'Twitter Ads', 'GT_Backend', 'Hootsuite Inc.',
       'Twitter Media Studio', 'aa.com.tr', 'Sprout Social',
       'Twitter VIT App for iOS', 'Facelift-Cloud', 'Instagram',
       'HubSpot', 'China Xinhua News', 'xh_scitech', 'Sprinklr Publisher',
       'Hearsay Social', 'Sprinklr Publishing',
       'Salesforce - Social Studio', 'Falcon Social Media Management ',
       'Castanet', 'Tweetbot for iΟS', 'cmssocialservice', 'Paper.li',
       'Sprinklr', 'Libsyn On-Publish', 'Twitter for iPad',
       'Premium Times Main', 'WordPress.com', 'SocialReport.com',
       'QuickTake by Bloomberg', 'LinkedIn', 'Khoros',
       'YabaLeftOnline Media', 'Grabyo', 'OneCMS Social Connect',
       'AgoraPulse Manager',

In [26]:
def agrupar_dispositivo(cliente):
    cliente_min = str(cliente).lower()
    if 'iphone' in cliente_min or 'ipad' in cliente_min:
        return 'Apple iOS'
    elif 'android' in cliente_min:
        return 'Android'
    elif 'web' in cliente_min:
        return 'Web (Escritorio)'
    elif 'tweetdeck' in cliente_min or 'hootsuite' in cliente_min:
        return 'Herramientas Profesionales'
    else:
        return 'Otros / Bots'

df['Dispositivo_Limpio'] = df['Client'].apply(agrupar_dispositivo)

In [27]:
df.to_csv("COVID_Clean.csv", index=False)

print("¡Fase completada con éxito! Archivo 'COVID_Clean.csv' generado.")

¡Fase completada con éxito! Archivo 'COVID_Clean.csv' generado.
